### PR TITLE
Editor: Request post selector posts by rows rendered

### PR DESCRIPTION
**Pull request by aduth**
This pull request seeks to enhance the infinite loading of the `<PostSelector />` component by triggering network requests specifically corresponding to the indices of the rows rendered in the selector. This is a prerequisite for state persistence of posts query data, as otherwise it is not possible to reliably or responsibly refresh persisted query state. State persistence of post queries will assist in faster initial load experiences of not only the `<PostSelector />` component, but also upcoming custom post types listing screens.

__Testing instructions:__

There are no visual or usability changes by the changes in this pull request. Rather, ensure that there are no regressions to existing behavior.

Verify that the post selector works in both "Add Link" existing content, and in page editor parent page selector. Ensure that compact view is shown when fewer than 7 posts exist, and that pages are loaded infinitely (with indicator) until all pages are exhausted. Verify that search continues to behave as expected, including pagination within search and clearing search results.

Test live: https://calypso.live/?branch=update/post-selector-rows-rendered

### CODE REVIEW

File: client/my-sites/post-selector/selector.jsx [hunk 1]
```diff
@@ -44,21 +49,21 @@ const PostSelectorPosts = React.createClass( {
 		posts: PropTypes.array,
 		postsHierarchy: PropTypes.array,
 		page: PropTypes.number,
-		lastPage: PropTypes.bool,
+		lastPage: PropTypes.number,
```
Comment (aduth) [inline client/my-sites/post-selector/selector.jsx]: @dmsnell : At this point, I might admit there can be confusion around naming of props without the `is` prefix 😄  (previously indicating equivalent of `isLastPage`, now the actual page number).

**Comment by aduth** [comment]
cc @timmyc some overlap with `<TermTreeSelector />`, couple general improvements (i.e. using `AutoSizer` instead of `getResultsWidth` function).

File: client/my-sites/post-selector/selector.jsx [hunk 2]
```diff
@@ -85,24 +89,31 @@ const PostSelectorPosts = React.createClass( {
 		}, SEARCH_DEBOUNCE_TIME_MS );
 	},
 
+	componentWillReceiveProps( nextProps ) {
+		if ( ! isEqual( this.props.query, nextProps.query ) ) {
+			this.setState( {
+				requestedPages: [ 1 ]
```
Comment (timmyc) [inline client/my-sites/post-selector/selector.jsx]: does `searchTerm` also need to be reset here?

File: client/my-sites/post-selector/selector.jsx [hunk 3]
```diff
@@ -340,48 +358,50 @@ const PostSelectorPosts = React.createClass( {
...
-							rowCount={ rowCount }
-							estimatedRowSize={ ITEM_HEIGHT }
-							rowHeight={ this.getRowHeight }
-							rowRenderer={ this.renderRow }
-							className="post-selector__results" />
-					) }
-				</InfiniteLoader>
+				<div className="post-selector__results">
+					<AutoSizer
+						key={ JSON.stringify( query ) }
```
Comment (timmyc) [inline client/my-sites/post-selector/selector.jsx]: Just curious why you need to set a key here?

File: client/state/posts/utils.js [hunk 4]
```diff
@@ -29,7 +35,7 @@ export function getNormalizedPostsQuery( query ) {
  */
 export function getSerializedPostsQuery( query = {}, siteId ) {
 	const normalizedQuery = getNormalizedPostsQuery( query );
-	const serializedQuery = JSON.stringify( normalizedQuery ).toLocaleLowerCase();
+	const serializedQuery = JSON.stringify( normalizedQuery );
```
Comment (timmyc) [inline client/state/posts/utils.js]: Why the change here?

**Comment by timmyc** [comment]
Testing out well for me.  Any idea why a request to `users` is fired off after each posts payload lands?

File: client/my-sites/post-selector/selector.jsx (hunk 2 above)
Comment (aduth) [inline client/my-sites/post-selector/selector.jsx]: >does `searchTerm` also need to be reset here?

`search` is a prop passed in from the parent component and, for that matter, could be the reason that `query` changed.

File: client/my-sites/post-selector/selector.jsx (hunk 3 above)
Comment (aduth) [inline client/my-sites/post-selector/selector.jsx]: >Just curious why you need to set a key here?

As in #6427, there's a few gotchas with `ReactVirtualized`, specifically when the row count doesn't change. The idea with adding `key` here is to address cases such as where a query changes but the resulting `postsHierarchy` results in the `rowCount` remaining the same (in fact, this might be an issue in production currently).

File: client/state/posts/utils.js (hunk 4 above)
Comment (aduth) [inline client/state/posts/utils.js]: >Why the change here?
 

I added an extended description to 988ea7d26d6c4bf62efe894ba9719182f6656c87 which explains why `toLowerCase` is a nice convenience for deduping, but in this case prevents accurately deserializing the key.

Related: `QueryManagerKey` [does not lowercase the key](https://github.com/Automattic/wp-calypso/blob/8238571290db36b15e986fd04c83c3311708d8b0/client/lib/query-manager/key.js#L20), and hasn't been a problem.

File: client/my-sites/post-selector/selector.jsx (hunk 2 above)
Comment (timmyc) [inline client/my-sites/post-selector/selector.jsx]: Oh, the reason I ask is it is being set in `getInitialState`

Comment (timmyc) [inline client/my-sites/post-selector/selector.jsx]: Based upon the logic in `onSearch` here it looks like `this.state.searchTerm` _should_ always be the same as the value coming in via the query.  What about adding a check for equality of `this.props.siteId !== nextProps.siteId`

**Comment by timmyc** [comment]
This is testing out well for me, just the one item about checking for `siteId` equality, but otherwise good to merge.

File: client/my-sites/post-selector/selector.jsx (hunk 2 above)
Comment (aduth) [inline client/my-sites/post-selector/selector.jsx]: >Oh, the reason I ask is it is being set in `getInitialState`

Oh, right, forgot about that. Hmm, this seems a little fragile / redundant, wonder if we can clean it up so we're relying exclusively on the incoming query prop. I imagine it's there to keep the UI "live" with the search entry, while only triggering the search request after a debounce.

>What about adding a check for equality of `this.props.siteId !== nextProps.siteId`

For sure, makes sense.

Comment (aduth) [inline client/my-sites/post-selector/selector.jsx]: Added `siteId` check in rebased 0e4ec6d. The search refactoring seems a bit more involved so is probably best suited for a separate pull request.
